### PR TITLE
[Sparkler 10] Extract text/title and push to crawldb

### LINKED ISSUE #10: Integrate Tika Parser to Parse Function for extracting Text and metadata
The current parse function is limited to outlinks

### PR DESCRIPTION
**Pull request by karanjeets**
Closes #10 

@thammegowda : Please review. Do you think it is the right time I should pull all the keys from Tika's Metadata and push it to Solr like we did in parserIndexer? As of now, only title field is extracted.

### CODE REVIEW

File: sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/ParseFunction.scala [hunk 1]
```diff
@@ -17,37 +17,49 @@
...
-import org.apache.tika.sax.LinkContentHandler
+import org.apache.tika.sax.{WriteOutContentHandler, BodyContentHandler, LinkContentHandler}
 
 import scala.collection.JavaConverters._
 
 /**
   * Created by thammegr on 6/7/16.
   */
-object ParseFunction extends ((CrawlData) => Set[String]) with Serializable with Loggable {
+object ParseFunction extends ((CrawlData) => (String, Set[String], Metadata)) with Serializable with Loggable {
```
Comment (thammegowda) [inline sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/ParseFunction.scala]: @karanjeets 
Lets refine the return type of this function.
How about creating a data type(POJO class) that include all these three as attributes?

### Suggestion:
Define a datatype
```scala
class ParserData (title:String, links:Set[String], meta: Metadata) {}
```
Then, modify function signature to this:
```scala
ParseFunction extends ((CrawlData) => (ParseData)) with Serializable with Loggable {...}
```

File: sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/ParseFunction.scala [hunk 2]
```diff
@@ -17,37 +17,49 @@
...
+
+      // Parse Outlinks
       meta.set("resourceName", data.content.url)
       parser.parse(stream, linkHandler, meta)
-      stream.close()
-      linkHandler.getLinks.asScala.map(_.getUri.trim).filter(!_.isEmpty).toSet
+      val outlinks = linkHandler.getLinks.asScala.map(_.getUri.trim).filter(!_.isEmpty).toSet
+
+      // Parse Text
+      stream = new ByteArrayInputStream(data.content.content)
```
Comment (thammegowda) [inline sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/ParseFunction.scala]: Couple of things:
1. Please close the previous stream before opening a new stream
2. Check if  it is possible to get both the outlinks as well as plain text in single pass?

**Comment by thammegowda** [comment]
> Do you think it is the right time I should pull all the keys from Tika's Metadata and push it to Solr like we did in parserIndexer?

Yes, lets get it done. Thanks

File: sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/ParseFunction.scala (hunk 1 above)
Comment (karanjeets) [inline sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/ParseFunction.scala]: ACK. I will change it. This looks cleaner.

File: sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/ParseFunction.scala (hunk 2 above)
Comment (karanjeets) [inline sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/ParseFunction.scala]: 1. ACK.
2. I checked but there isn't any handler which does both things at once. We might need to do this one after the another. Do you have any suggestion?

**Comment by karanjeets** [comment]
> Yes, lets get it done. Thanks

Sure, I will add the changes soon.

**Comment by karanjeets** [comment]
@thammegowda : Changes have been made. Please review.

File: conf/solr/crawldb/conf/managed-schema [hunk 3]
```diff
@@ -120,8 +120,21 @@
...
+   <dynamicField name="*_t_md" type="text_general" indexed="true" stored="true"/>
+   <dynamicField name="*_ts_md" type="text_general" indexed="true" stored="true" multiValued="true"/>
+   <dynamicField name="*_i_md" type="int" indexed="true" stored="true"/>
+   <dynamicField name="*_is_md" type="int" indexed="true" stored="true" multiValued="true"/>
+   <dynamicField name="*_l_md" type="long" indexed="true" stored="true"/>
+   <dynamicField name="*_ls_md" type="long" indexed="true" stored="true" multiValued="true"/>
+   <dynamicField name="*_d_md" type="double" indexed="true" stored="true"/>
+   <dynamicField name="*_ds_md" type="double" indexed="true" stored="true" multiValued="true"/>
+   <dynamicField name="*_b_md" type="boolean" indexed="true" stored="true"/>
+   <dynamicField name="*_bs_md" type="boolean" indexed="true" stored="true" multiValued="true"/>
```
Comment (thammegowda) [inline conf/solr/crawldb/conf/managed-schema]: Please setup the copy fields properly.
The best practice is to have a field named `text` which is indexed but not stored.
We need to copy all these fields to the `text` field. Checkout [these lines from](https://github.com/USCDataScience/parser-indexer/blob/0418e8f2f2589117e90f03b52508b7238c2894be/conf/schema.xml#L504-L513) parser-indexer for  reference.
I hope you know this, but missed to do it in hurry. Ask me if you are unsure why we do that way.

File: sparkler-api/src/main/java/edu/usc/irds/sparkler/solr/schema/StringEvaluator.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+    public static String BOOL_REGEX = "(?i)^(true|false)$";
+    public static String DOUBLE_REGEX = "^[+-]?\\d+(\\.\\d+)?([Ee][+-]\\d+)?$";
+    private static LinkedHashMap<String, Eval> evals = new LinkedHashMap<>();
+    static {
+        //regex  -> evaluator
+        //evals.put(INT_REGEX, Integer::parseInt); //Let the long take it
+        evals.put(LONG_REGEX, Long::parseLong);
+        evals.put(BOOL_REGEX, Boolean::parseBoolean);
+        evals.put(DOUBLE_REGEX, Double::parseDouble);
```
Comment (thammegowda) [inline sparkler-api/src/main/java/edu/usc/irds/sparkler/solr/schema/StringEvaluator.java]: Nicely Done ;-)

Comment (karanjeets) [inline sparkler-api/src/main/java/edu/usc/irds/sparkler/solr/schema/StringEvaluator.java]: Haha. You know the secret. :)
(Taken from parser-indexer)

File: conf/solr/crawldb/conf/managed-schema (hunk 3 above)
Comment (karanjeets) [inline conf/solr/crawldb/conf/managed-schema]: Ohh, yes. I missed it. Yeah, I know the reason for this. 
I will add them and make another commit.

Comment (karanjeets) [inline conf/solr/crawldb/conf/managed-schema]: @thammegowda : Done. Run a complete test crawl and let me know if any issues.

**Comment by thammegowda** [comment]
@karanjeets 

Got this issue in the crawldb. Working on it.
<pre>
Caused by: java.lang.IllegalArgumentException: Document contains at least one immense term in field="plainText" (whose UTF8 encoding is longer than the max length 32766), all of which were skipped.  Please correct the analyzer to not produce such terms.  The prefix of the first immense term is: '[10, 32, 32, 32, 32, 10, 10, 32, 32, 32, 32, 10, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 32, 10, 10, 10, 10, 32]...', original message: bytes can be at most 32766 in length; got 43197
	at org.apache.lucene.index.DefaultIndexingChain$PerField.invert(DefaultIndexingChain.java:772)
	at org.apache.lucene.index.DefaultIndexingChain.processField(DefaultIndexingChain.java:417)
	at org.apache.lucene.index.DefaultIndexingChain.processDocument(DefaultIndexingChain.java:373)
	at org.apache.lucene.index.DocumentsWriterPerThread.updateDocument(DocumentsWriterPerThread.java:232)
	at org.apache.lucene.index.DocumentsWriter.updateDocument(DocumentsWriter.java:449)
	at org.apache.lucene.index.IndexWriter.updateDocument(IndexWriter.java:1492)
	at org.apache.solr.update.DirectUpdateHandler2.doNormalUpdate(DirectUpdateHandler2.java:282)
	at org.apache.solr.update.DirectUpdateHandler2.addDoc0(DirectUpdateHandler2.java:214)
	at org.apache.solr.update.DirectUpdateHandler2.addDoc(DirectUpdateHandler2.java:169)
	... 48 more
Caused by: org.apache.lucene.util.BytesRefHash$MaxBytesLengthExceededException: bytes can be at most 32766 in length; got 43197
	at org.apache.lucene.util.BytesRefHash.add(BytesRefHash.java:280)
	at org.apache.lucene.index.TermsHashPerField.add(TermsHashPerField.java:150)
	at org.apache.lucene.index.DefaultIndexingChain$PerField.invert(DefaultIndexingChain.java:762)
	... 56 more

</pre>

**Comment by thammegowda** [comment]
@karanjeets 
FYI,
```xml
 <field name="plainText" type="string" indexed="true" 
      stored="true" multiValued="false" docValues="true" />
```
String is for exact match (or to keep untokenized for regex match), `text_general` is the best fit here

File: sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/ParseFunction.scala [hunk 5]
```diff
@@ -17,37 +17,57 @@
...
+        LOG.warn("PARSING-OUTLINKS-ERROR {}", data.content.url)
         LOG.warn(e.getMessage, e)
-        Set.empty[String]
+    } finally {
+      IOUtils.closeQuietly(stream)
+    }
+    try {
+      // Parse Text
+      stream = new ByteArrayInputStream(data.content.content)
+      parser.parse(stream, contentHandler, meta)
```
Comment (thammegowda) [inline sparkler-app/src/main/scala/edu/usc/irds/sparkler/pipeline/ParseFunction.scala]: @karanjeets 
FYI reusing `meta` object for two parse calls causes duplicate key:value insertions!
Fixing it.

**Comment by thammegowda** [comment]
@karanjeets Tested, fixed few issues and merged. Thanks :+1:

**Comment by karanjeets** [comment]
Great 👍

**Pull request merged**